va/va_prot.h

```diff
@@ -0,0 +1,361 @@
+ * tee_exec_buff.output.max_data_size = (uint16_t);
+ * va_status = vaCreateBuffer(
+ *                 va_dpy,
+ *                 protected_session,
+ *                 (VABufferType) VAProtectedSessionTeeExecBufferType,
+ *                 sizeof(tee_exec_buff),
+ *                 1,
+ *                 &tee_exec_buff,
+ *                 &buffer);
+ *
+ * va_status = vaProtectedSessionTeeExec(va_dpy, protected_session, buffer);
+ *
+ * vaDestroyBuffer(va_dpy, buffer);
+ * \endcode
+ */
+
+/**
+ * ProtectedSessions and Contexts
+ *
+ * Context represents a "virtual" video decode, encode or video processing
+ * pipeline. Surfaces are render targets for a given context. The data in the
+ * surfaces are not accessible to the client except if derived image is supported
+ * and the internal data format of the surface is implementation specific.
+ *
+ * A protected session is a session between application and GPU hardware.
+ * Protected sessions can be attached to display to enable protection in the pipeline.
+ *
+ * Both contexts and protected sessions are identified by unique IDs and its
+ * implementation specific internals are kept opaque to the clients
+ */
+typedef VAContextID VAProtectedSessionID;
+
+/** \brief Input buffer of VAProtectedSessionTeeExecBuffer or VAProtectedSessionHwUpdateBuffer */
+typedef struct _VAProtectedFuncionBufferInput
+{
+    /** \brief Contains the size of the input data. */
+    uint32_t    data_size;
+    /** \brief The input data. */
+    void        *input_data;
+} VAProtectedFuncionBufferInput;
+
+/** \brief Output buffer of VAProtectedSessionTeeExecBuffer or VAProtectedSessionHwUpdateBuffer */
+typedef struct _VAProtectedFuncionBufferOutput
+{
+    /** \brief The maximum size of data that the driver can return in the output buffer. */
+    uint32_t    max_data_size;
+    /** \brief Returns the size of the output data written by the driver. */
+    uint32_t    data_size;
+    /** \brief Output data written by the driver. */
+    void        *output_data;
+} VAProtectedFuncionBufferOutput;
+
+/** \brief Buffer for vaProtectedSessionTeeExec() */
+typedef struct _VAProtectedSessionTeeExecBuffer {
+    /** \brief Specify the function to execute. It is IHV's implementation specific */
+    uint32_t                        function_id;
+    /** \brief Input buffer of VAProtectedFuncionBufferInput structure */
+    VAProtectedFuncionBufferInput   input;
+    /** \brief Output buffer of VAProtectedFuncionBufferOutput structure */
+    VAProtectedFuncionBufferOutput  output;
+    /** \brief Return the result of this function. It is IHV's implementation specific */
+    uint32_t                        status;
+    uint32_t                        va_reserved[VA_PADDING_LOW];
+} VAProtectedSessionTeeExecBuffer;
+
+/** \brief Buffer for vaProtectedSessionHwUpdate() */
+typedef struct _VAProtectedSessionHwUpdateBuffer {
+    /** \brief Input buffer of VAProtectedFuncionBufferInput structure */
```

Comment:
You can just have function_id=HW_UPDATE and avoid definition of VAProtectedSessionHwUpdateBuffer which is 99% identical to the VAProtectedSessionTeeExecBuffer definition.

Reply:
These 2 functions are different from design. VAProtectedSessionExecute is to execute or retrieve from GPU/TEE while VAProtectedSessionHwUpdate is to update the status of protected session. In VAProtectedSessionHwUpdateBuffer, there is no function_id.